### PR TITLE
synchronize map layer access for NewMap

### LINKED ISSUE #7034: ArrayIndexOutOfBoundsException
From Google Play for beta version 2018.05.28-RC (but also a some reports for 2018.02.05):
```
java.lang.ArrayIndexOutOfBoundsException: 
 
  at java.util.concurrent.CopyOnWriteArrayList.get (CopyOnWriteArrayList.java:121)
 
  at org.mapsforge.map.layer.Layers.get (Layers.java:230)
 
  at org.mapsforge.map.android.input.TouchGestureHandler.onSingleTapConfirmed (TouchGestureHandler.java:233)
 
  at android.view.GestureDetector$GestureHandler.handleMessage (GestureDetector.java:303)
 
  at android.os.Handler.dispatchMessage (Handler.java:105)
 
  at android.os.Looper.loop (Looper.java:156)
 
  at android.app.ActivityThread.main (ActivityThread.java:6523)
 
  at java.lang.reflect.Method.invoke (Native Method)
 
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:942)
 
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:832)
```

### LINKED ISSUE #8126: [Nightly] Live Map crash with OSM and OpenAndroMaps
**Describe the bug:**
When Live Map shows a reasonable amount of caches, moving the map around crashes the map.

**To Reproduce:**
1. open e. g. GC7NDVZ
2. open map view by e. g. clicking listing page's compass rose
3. select an offline map (I tested OSM and two different OpenAndroMaps)
4. zoom out to a scale of 2 km
5. activate "update map" (as for me I see 364 caches right now)
6. move the map around a bit ... (500 m west then 1000 back east is sufficient over here)
7. crash ... then listing page will be shown

**Version of c:geo used:**
2020.02.14-NB-7d86095

**Is the problem reproducible:**
Yes

**System information:**
```
--- System information ---
Device: VOG-L29 (VOG-L29EEA, HUAWEI)
Android version: 10
Android build: VOG-L29 10.0.0.178(C431E19R2P5)
c:geo version: 2020.02.14-NB-7d86095
Google Play services: disabled - 20.04.14 (120400-294335909)
Low power mode: inactive
Compass capabilities: yes
Rotation vector sensor: present
Orientation sensor: present
Magnetometer & Accelerometer sensor: present
Direction sensor used: rotation vector
Hide own/found: true
HW acceleration: enabled (default state)
System language: de_DE
System date format: dd.MM.yy
Debug mode active: no
System internal c:geo dir: /data/user/0/cgeo.geocaching (24,4 GB free) internal
User storage c:geo dir: /storage/emulated/0/cgeo (24,4 GB free) external non-removable
Geocache data: /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData (24,4 GB free) external non-removable
Database: /data/user/0/cgeo.geocaching/databases/data (263,9 MB) on system internal storage
Fine location permission: granted
Write external storage permission: granted
Geocaching sites enabled:
   geocaching.com: Logged in (Anmeldung OK) / PREMIUM
   extremcaching.com: Logged in (Anmeldung OK)
Geocaching.com date format: dd MMM yy
Installed c:geo plugins: contacts
--- End of system information ---
```

**Additional context:**
Maps used are current OSM (online) and OpenAndroMaps "Germany_South" and "Baden-Wuerttemberg" (offline) as of June 2019, as well as current BaWü, freshly downloaded at http://download.openandromaps.org/mapsV4/Germany/baden-wuerttemberg.zip (as of 29/01/2020) - the latter takes some more moves, but crashes at the end, too. The more caches are displayed, the sooner the map crashes.

### PR DESCRIPTION
**Pull request by moving-bits**
proposal to fix #7034 and fix #8126

Do not merge for now, but field test first. Please look for possible side effects as well.

For details see https://github.com/cgeo/cgeo/issues/8126#issuecomment-605324757

### CODE REVIEW

**Comment by Lineflyer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ main/src/cgeo/geocaching/maps/mapsforge/v6/MfMapView.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/cgeo/cgeo/pullRequest?prid=5261366&bid=17237192)

**Comment by moving-bits** [comment]
@MagpieFourtyTwo
@jakedot
@Lineflyer 
As you've encountered the underlying issue #8126 on your physical devices - do you have a chance to field test to see if this PR fixes the problem, and to watch for possible side effects?

@Lineflyer 
Or shall we merge it to master first and test it in the nightlies?

**Comment by MagpieFourtyTwo** [comment]
> do you have a chance to field test to see if this PR fixes the problem
I of course would do so - but can't compile to apk on my own, as I don't have a (working) SDK right now.  Thus currently all I can do is using the nightlies, so the merge would be the easiest way for me.  Or you perhaps may deliver an apk (which will not break the version timeline, so afterwards I can use "normal" nightlies again.

**Comment by bekuno** [comment]
@MagpieFourtyTwo 
If you follow the CI check then you find *.apk to download.
In this case at https://ci.cgeo.org/job/cgeo%20pull%20request/2582/.
The cgeo-basic-debug-androidTest.apk is enough without a dev environment.

**Comment by moving-bits** [comment]
Just make a backup of your database (and your program settings, if you like) before, using c:geo's backup features. This prepares for switching between master and release branch versions, and for uninstalling and reinstalling c:geo without data loss, if needed.

**Comment by moving-bits** [comment]
> The cgeo-basic-debug-androidTest.apk is enough without a dev environment.

cgeo-basic-debug-androidTest.apk or cgeo-debug.apk?

File: main/src/cgeo/geocaching/maps/mapsforge/v6/caches/CachesBundle.java [hunk 1]
```diff
@@ -298,6 +298,10 @@ Layer getCirclesSeparator() {
         return separators.get(CIRCLES_SEPARATOR);
     }
 
+    MfMapView getMapView() {
```
[the remark this answers is not in the document]
Comment (moving-bits) [reply main/src/cgeo/geocaching/maps/mapsforge/v6/caches/CachesBundle.java]: updated PR accordingly

**Comment by moving-bits** [comment]
Note: build fails due to errors in gradle config:

```
* What went wrong:
16:57:29  Could not compile settings file '/home/jenkins/slave/workspace/cgeo pull request/settings.gradle'.
16:57:29  > startup failed:
16:57:29    settings file '/home/jenkins/slave/workspace/cgeo pull request/settings.gradle': 4: Only Project build scripts can contain plugins {} blocks
16:57:29    
16:57:29    See https://docs.gradle.org/5.6.4/userguide/plugins.html#sec:plugins_block for information on the plugins {} block
16:57:29    
16:57:29     @ line 4, column 1.
16:57:29       plugins {
16:57:29       ^
16:57:29    
16:57:29    1 error
```

**Comment by Lineflyer** [comment]
Note: build fails due to errors in gradle config:

Yepp, my bad: See #8162

**Comment by bekuno** [comment]
retest this please

**Comment by moving-bits** [comment]
retest this please

**Comment by MagpieFourtyTwo** [comment]
@bekuno 
Sorry, but I lost you completely.  In fact I don't even know how I could have come to the mentioned CI 2582 on my own. Moreover I don't know what can be done with which apk listed there:

- cgeo-basic-debug-androidTest.apk - just seems to install a testing environment, which I don't know how to use
- cgeo-debug.apk - does not install at all (Android's extremely informative error message: "App has not been installed.")

So, of course I would like to help/test/whatever, but I'm afraid, I need a little bit more info on how to ...

**Comment by bekuno** [comment]
You should see something like this screenshot at the end of this PR:
grafik

There you have to click at "Details" for the "default" check.
I'm not sure if you would see this.

On the cgeo CI PR site
https://ci.cgeo.org/job/cgeo%20pull%20request/
you find at the left side the buils, e.g. for this PR 
grafik
You can click then on the build number for more details.

**Comment by bekuno** [comment]
>cgeo-debug.apk - does not install at all (Android's extremely informative error message: "App has not been installed.")

Yes, the message is not helpful.
Mostly the installation does not work when conflicting with the installed version. Please uninstall WhereYouGo first.

**Comment by moving-bits** [comment]
> Yes, the message is not helpful.
> Mostly the installation does not work when conflicting with the installed version. Please uninstall WhereYouGo first.

c:geo, not WhereYouGo

Please remember to backup database (and program settings) first, using c:geo's backup functions, then you can restore both after installing the apk.

**Comment by jakedot** [comment]
@moving-bits: do you know why the debug version doesn't recognize the saved
settings/database for restoring?

Without having my usual environment restored, I can't say for sure if the
problem is gone...

Am So., 29. März 2020 um 20:11 Uhr schrieb moving-bits <
notifications@github.com>:

**Comment by Lineflyer** [comment]
> @moving-bits: do you know why the debug version doesn't recognize the saved
> settings/database for restoring?
> 

Thats strange indeed.

Lets try to get the build status green, so we merge it for test in the nightlies.

**Comment by Lineflyer** [comment]
retest this please

**Comment by moving-bits** [comment]
> @moving-bits: do you know why the debug version doesn't recognize the saved settings/database for restoring?

I don't know for sure, but I've encountered it sometimes as well. It might have something todo with access rights. If c:geo hasn't asked you for storage access permissions yet (and therefore you could not grant it), it cannot find the backup file.
If that happens after doing a fresh install, I close c:geo, remove it from memory, and then restart it. c:geo then asks me for location and storage access permissions, and after granting at least the storage access permissions, it can access its backup files.

**Comment by Lineflyer** [comment]
@moving-bits Feel free to merge if you think its ready for that. I will also try to test in nightlies.

**Comment by moving-bits** [comment]
> @moving-bits Feel free to merge if you think its ready for that. I will also try to test in nightlies.

It's ready AFAICS, at least during my tests I haven't been able to crash c:geo any longer, even after 60 seconds of rapid eight-figures etc., whereas without this fix, it took me no longer than ten seconds. => I'll merge it.

**Comment by Lineflyer** [comment]
OK, I will build a new nightly right away.

BTW:
I will try to be more online on IRC from now on, so that the team might have higher chance to get in touch outside of the issue threads :)

**Comment by MagpieFourtyTwo** [comment]
@bekuno 
> You should see something like this screenshot at the end of this PR:
> grafik
> 
> There you have to click at "Details" for the "default" check.

Thanks for bringing some light into this, now I'm in sync again.  ;)
Ok, to get this clear: What I (would) need is not 'cgeo-basic-debug-androidTest.apk', as you first wrote (BTW: what is this thingy for?), but in fact 'cgeo-debug.apk' - right?

Nevertheless - as un/re/new-installing seems to potentially give new problems and as the merge will be done anyway, I just wait for the nightly.

Last question: How do I know if a published nightly already contains a (this) fix?

**Comment by moving-bits** [comment]
> Ok, to get this clear: What I (would) need is not 'cgeo-basic-debug-androidTest.apk', as you first wrote (BTW: what is this thingy for?), but in fact 'cgeo-debug.apk' - right?

Yes.
cgeo-basic-debug-androidTest.apk is the test package (used during development).

> Nevertheless - as un/re/new-installing seems to potentially give new problems and as the merge will be done anyway, I just wait for the nightly.

Totally understandable.
Anyway, if you ever encounter the need to install a debug apk, just follow the following steps to be on the safe side:
- backup your database (and settings, if needed), using c:geo's backup feature
- uninstall c:geo
- install the apk
- start c:geo and grant storage access rights
- restore the backup
(and if you want to go back to Play store version, just replace "the apk" with "c:geo from Play store" in step 3)

> Last question: How do I know if a published nightly already contains a (this) fix?

Nightlies are built from master branch, so every pull request that got merged into master branch before the date of your nightly version is included. See the "pulls" tab and filter for "merged".

**Comment by jakedot** [comment]
@MagpieFourtyTwo You know if this (or any other) fix is contained if the corresponding PR has been merged into the master branch, the nightly comes from that branch.

**Comment by Lineflyer** [comment]
To be very precise:
You can have a look into the commit history of the master branch (by selecting the code tab above and then selecting commits).
Each commit has an unique hash (listed on the right).

If you look into the "About c:geo" setion of the nightly build the version name contains the hash from which it has been build. This means everything including the commit with this hash (chronological) is included in the nightly.

**Comment by MagpieFourtyTwo** [comment]
Thanks a lot!

And, by the way ... if anybody cares ... THE ISSUE SEEMS TO BE SOLVED!!!  :)))

Installed 2020.03.29-NB2-2855b29 in the second it came out and since then I was drawing 8 figures on my (poor) screen ... with not a single crash.  Ok, apart from the screen protector, which needs to be replaced now, in order to get rid of the engraved 8s ...  ;))

Even tried different scenarios (Live Map from start page, activated Map Update from cache listing, from nearest and waypoint list) - and it really seems to be stable.

Great! Thanks a lot! This brings pleasure back. :)

**Comment by MagpieFourtyTwo** [comment]
As we're just talking about it:

> restore the backup (and if you want to go back to Play store version, just replace "the apk" with "c:geo from Play store" in step 3)

... tried this a year ago or so, but failed, as c:geo was not willing to accept a backup from a higher (i. e. the nightly's) version. Which sounded plausible to me ... has there something changed to make it possible to use a nightly's backup in the play store version?

**Comment by Lineflyer** [comment]
>  has there something changed to make it possible to use a nightly's backup in the play store version?

Nope. We do not have any downgrade conversion for the c:geo database, thus its not possible. Going back from nightly to Playstore version is not a really relevant scenario, besides for testing.

**Comment by moving-bits** [comment]
> ... tried this a year ago or so, but failed, as c:geo was not willing to accept a backup from a higher (i. e. the nightly's) version. Which sounded plausible to me ... has there something changed to make it possible to use a nightly's backup in the play store version?

This is not a general problem, it only happens when there is a change of the database structure involved. (It's not about program version that conflicts here, but about database version.)

If you created a backup from your Play store version and would like to go back to that, simply uninstall the dev / nightly version without creating a backup from dev db (which would overwrite the backup you created before installing the dev version), then reinstall the Play store version. - I'll update my comment above regarding going back to Play store version.

**Comment by MagpieFourtyTwo** [comment]
Thought so.

Ok, but then it should at least be possible, to restore a (newer) play store version's data with a (older) nightly's backup, which has been created prior to the play store version, right?

Background: Using nightlies for several years now and *if* there some day *may be* a reason to go back to the play store version, one would not like to lose all the data collected in these years ...

**Comment by moving-bits** [comment]
> Ok, but then it should at least be possible, to restore a (newer) play store version's data with a (older) nightly's backup, which has been created prior to the play store version, right?

Yes, in terms of database structure this should always be possible. (Right now we're at DB version 80 on master, IIRC, and one PR setting this to 81 (#8158), but we support upgrading from DB version 50 onward.)

**Comment by moving-bits** [comment]
> And, by the way ... if anybody cares ... THE ISSUE SEEMS TO BE SOLVED!!! :)))

I do ;-)
Good to hear it works. Thanks for testing!

> Installed 2020.03.29-NB2-2855b29 in the second it came out and since then I was drawing 8 figures on my (poor) screen ... with not a single crash. Ok, apart from the screen protector, which needs to be replaced now, in order to get rid of the engraved 8s ... ;))

He he. Maybe you should have switched between 8 und other figures more frequently ;-)

**Pull request merged**